### DISCUSSION TITLE
Save MobX state in LocalStorage

### DISCUSSION THREAD

**Discussion by deadcoder0904** [mobxjs/mobx#2798] [Q&A] [answered]
Originally, I had my store working when I was using `new CounterStore` inside `React.createContext()`

# context.ts

```tsx
import React from 'react'
import { stores, PersistState, CounterStore } from '@/store/index'
import type { ICounterStore } from '@/types/index'

export const FrameItContext = React.createContext<ICounterStore>(new CounterStore())
export const useCounterStore = () => React.useContext(FrameItContext)
```

Then I started using [Mobx Persist Store](https://github.com/quarrant/mobx-persist-store/) in my app.

# persist.ts

```tsx
import { persistence, StorageAdapter } from 'mobx-persist-store'
import { CounterStore } from '@/store/index'

const read = (name: string): Promise<string> =>
	new Promise((resolve) => {
		const data = localStorage.getItem(name) || '{}'
		resolve(data)
	})

const write = (name: string, content: string): Promise<Error | undefined> =>
	new Promise((resolve) => {
		localStorage.setItem(name, JSON.stringify(content))
		resolve(undefined)
	})

export const PersistState = persistence({
	name: 'CounterStore',
	properties: ['counter', 'increment', 'decrement'],
	adapter: new StorageAdapter({ read, write }),
	reactionOptions: {
		// optional
		delay: 2000,
	},
})(new CounterStore())
```

And I changed my code to use `PersistState` instead of `new CounterStore()` & it stopped working.

# context.ts

```tsx
import React from 'react'
import { stores, PersistState, CounterStore } from '@/store/index'
import type { ICounterStore } from '@/types/index'

export const FrameItContext = React.createContext<ICounterStore>(PersistState)
export const useCounterStore = () => React.useContext(FrameItContext)
```

How do I solve it?

**Comment by danielkcz** [comment]
You should elaborate on "it stopped working".

**Reply by deadcoder0904** [reply]
Oh sorry about that. What I mean by that is the UI shows empty screen & there are no errors in the console or terminal when I use `PersistState`.

And when I use `stores.frameItStore` where `export const stores = { frameItStore: PersistState }` instead of `PersistState`, then it gives an error in the console:

> Uncaught TypeError: Cannot read property 'frameitStore' of undefined

I even tried making `PersistState` a function which I then call in context like `React.createContext<ICounterStore>(PersistState())` but it again gives the same error as above.

**Reply by deadcoder0904** [reply]
Just a correction: I had `if (!isSynchronized(stores.frameitStore)) return null` in my top-level `App.tsx` but yeah I removed it now & it's not saving state to local storage when I use `PersistState` & it shows the `undefined` error above when I use `stores.frameItStore`.

So my question is how to make context work with Mobx Persist Store (MPS)? I have similar code as [the Counter example](https://codesandbox.io/s/fervent-sound-m7ie5?file=/src/App.tsx) that uses MPS but unable to convert it to use context :)

**Reply by danielkcz** [reply]
Yea, without runnable reproduction it's hard to say. Your snippets are not that useful because often the problem lies in things that are not present in those snippets.

**Reply by deadcoder0904** [reply]
Weird, it works in Codesandbox for a Counter example 🤦‍♂️ → https://codesandbox.io/s/mobx-persist-store-4l1dm

Which is kinda similar to the snippet I provided.

**Reply by danielkcz** [reply]
That's what I thought. You need to investigate your environment and compare the differences. I am pretty sure you will find out your answer on the way.

**Reply by deadcoder0904** [reply]
@FredyC I can't figure it out at all. The code is basically same. The `write` function in `StorageAdapter` never gets called for some reason.

Is there any way to persist individual elements in LocalStorage? Any examples/blogs are appreciated. The only npm library that supports mobx v6 doesn't work for me & stackoverflow doesn't help much.

I tried doing the following:

![mobx-store-image](https://user-images.githubusercontent.com/16436270/108964187-2c4d2e80-76a1-11eb-93ed-5771ef1c5a36.png)

It works but it saves everything from the store. I want to only save certain properties. Let me know if you have a solution :)

**Reply by danielkcz** [reply]
If you want to save yourself some pain, you can consider mobx-state-tree or mobx-keystone which offer persistence out-of-box. I don't have any experience persisting pure mobx, so good luck I guess :)

**Reply by urugator** [reply]
> It works but it saves everything from the store.

Just implement `toJSON` on store and use `JSON.stringify(_this)` ( you don't have to call `toJS`)

**Reply by deadcoder0904** [reply]
@urugator Not sure where do I have to do `toJSON`? I currently did → https://github.com/mobxjs/mobx/discussions/2798#discussioncomment-400780

**Reply by urugator** [reply]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior
```javascript
	constructor() {						
		makeObservable(this, {
			counter: observable,
			color: observable,
			increment: action.bound,
			decrement: action.bound,
			switchColor: action.bound,
		})
		Object.assign(this, JSON.parse(localStorage.getItem('foo')))
		autorun(() => {			
			localStorage.setItem('foo', JSON.stringify(this))
		})	
	}

	toJSON() {
		return {
			counter: this.counter,
			color: this.color,
		}
	}
```

**Reply by deadcoder0904** [reply]
Oh yeah, I know that already ahaha having read YDKJS but thanks for reminding :)

**Reply by deadcoder0904** [reply]
@urugator this is even more simple haha 😄

**Reply by deadcoder0904** [reply]
@urugator How does this work with multiple stores? Do I have to implement `toJSON` everywhere?

Or I can call `Auth.email` & `User.name` in `Root` store itself where `Auth` & `User` are there own MobX Stores & I call `toJSON()` in `Root`?

**Reply by urugator** [reply]
Why don't you try 😉

**Reply by deadcoder0904** [reply]
Yeah, I will when I separate my stores. But I think it works after reading your message 😜

**Comment by mweststrate** [comment]
Or just write it by hand: create a computed property that gives back the
things that you want to serialize. In general I don't think libraries for
this are in many cases not needed :)

**Reply by deadcoder0904** [reply]
I did what you told me but how do I get the current LocalStorage value when I reload the app? It stores perfectly in the LocalStorage now:

```tsx
private state = ""
counter = 0
color: Color = "red"

constructor() {
	makeObservable(this, {
		counter: observable,
		color: observable,
		increment: action.bound,
		decrement: action.bound,
		switchColor: action.bound,
	})

	autorun(() => {
		const name = "CounterStore"
		const properties = ["counter", "color"]
		const storedJson = localStorage.getItem(name)
		console.log(storedJson)
		if (storedJson) {
			this.state = JSON.parse(storedJson)
		}
		autorun(() => {
			// const value = toJS(_this)
			localStorage.setItem(name, JSON.stringify(pick(this, properties)))
		})
	})
}
```

Codesandbox → https://codesandbox.io/s/mobx-persist-store-manual-implementation-vm38r?file=/src/store.ts

**Comment by mweststrate** [comment]
Or to simplify it a bit: lodash.pick might get you pretty far

**Comment by deadcoder0904** [comment, marked as answer]
This is the code that works. `toJSON` contains the properties that need to be saved. Let me know if anything can be improved :)

Here's the Codesandbox → https://codesandbox.io/s/mobx-persist-store-manual-implementation-vm38r?file=/src/store.ts

```tsx
import { observable, action, makeObservable, computed, autorun } from "mobx"
import pick from "lodash.pick"

type Color = "red" | "blue" | "green"

export interface ICounterStore {
	counter: number
	color: Color
	switchColor(color: Color): void
	increment(n: number): void
	decrement(n: number): void
}

export class CounterStore implements ICounterStore {
	// private state = ""
	counter = 0
	color: Color = "red"

	constructor() {
		makeObservable(this, {
			counter: observable,
			color: observable,
			increment: action.bound,
			decrement: action.bound,
			switchColor: action.bound,
		})

		const name = "CounterStore"
		const storedJson = localStorage.getItem(name)
		if (storedJson) Object.assign(this, JSON.parse(storedJson))
		autorun(() => {
			localStorage.setItem(name, JSON.stringify(this))

			// const properties = ["counter", "color"]
			// const storedJson = localStorage.getItem(name)
			// console.log(storedJson)
			// if (storedJson) {
			// 	this.state = JSON.parse(storedJson)
			// }
			// // const value = toJS(_this)
			// localStorage.setItem(name, JSON.stringify(pick(this, properties)))
		})
	}

	increment(n: number) {
		this.counter = this.counter + n
	}

	decrement(n: number) {
		this.counter = this.counter - n
	}

	switchColor(color: Color) {
		if (color === "red") this.color = "blue"
		else if (color === "blue") this.color = "green"
		else this.color = "red"
	}

	toJSON() {
		return {
			color: this.color,
			counter: this.counter,
		}
	}
}

export const store = new CounterStore()
```

Thank you @mweststrate @urugator
